src/data/list/basic.lean

```diff
@@ -2975,6 +2978,40 @@ instance decidable_suffix [decidable_eq α] : ∀ (l₁ l₂ : list α), decidab
   | s, t, ⟨b::l, he⟩ := list.no_confusion he (λab lt, or.inr ⟨l, lt⟩)
   end⟩
 
+lemma inits_cons (a : α) (l : list α) : inits (a :: l) = [] :: l.inits.map (λ t, a :: t) :=
+by simp
+
+lemma tails_cons (a : α) (l : list α) : tails (a :: l) = (a :: l) :: l.tails :=
+by simp
+
+@[simp]
+lemma inits_append : ∀ (s t : list α), inits (s ++ t) = s.inits ++ t.inits.tail.map (λ l, s ++ l)
+| [] [] := by simp
+| [] (a::t) := by simp
+| (a::s) t := by simp [inits_append s t]
+
+@[simp]
+lemma tails_append : ∀ (s t : list α), tails (s ++ t) = s.tails.map (λ l, l ++ t) ++ t.tails.tail
+| [] [] := by simp
+| [] (a::t) := by simp
+| (a::s) t := by simp [tails_append s t]
+
+lemma reverse_map_reverse_tails_reverse :
+  ∀ (l : list α), reverse (map reverse $ tails $ reverse l) = l.inits
+| [] := by simp
+| (a :: l) := by simp [← reverse_map_reverse_tails_reverse l, map_eq_map_iff]
+
+lemma reverse_map_reverse_inits_reverse :
+  ∀ (l : list α), reverse (map reverse $ inits $ reverse l) = l.tails
+| [] := by simp
+| (a :: l) := by simp [← reverse_map_reverse_inits_reverse l, append_left_inj]
+
+lemma inits_reverse (l : list α) : inits (reverse l) = map reverse (reverse l.tails) :=
+by { rw ← reverse_map_reverse_inits_reverse l, simp [reverse_involutive.comp_self], }
+
+lemma tails_reverse (l : list α) : tails (reverse l) = map reverse (reverse l.inits) :=
+by { rw ← reverse_map_reverse_tails_reverse l, simp [reverse_involutive.comp_self], }
```

Comment:
I'm not sure whether all four of these are sensible, or if I should just keep one pair. `reverse ∘ map reverse` is involutive, so it's a question of whether it should appear on the left or the right of the lemma.

Reply:
It makes sense to keep both: sometimes you want to talk about `inits l` but you only have `tails`, and sometimes you want to talk about `inits (reverse l)`.

Reply:
The problem with that logic is that there is an explosion of possible orderings, all of which are trivial to obtain from the others (`x` marks the ones in this PR):
* [x] `tails = reverse ∘ map reverse ∘ inits ∘ reverse`
* [ ] `tails = map reverse ∘ reverse ∘ inits ∘ reverse`
* [ ] `tails ∘ reverse = reverse ∘ map reverse ∘ inits`
* [x] `tails ∘ reverse = map reverse ∘ reverse ∘ inits`
* [ ] `map reverse ∘ tails = reverse ∘ inits ∘ reverse`
* [ ] `reverse ∘ map reverse ∘ tails = inits ∘ reverse`
* [x] `map reverse ∘ reverse ∘ tails = inits ∘ reverse`
* [x] `reverse ∘ map reverse ∘ tails ∘ reverse = inits`
* [ ] `map reverse ∘ reverse ∘ tails ∘ reverse = inits`

All of these follow trivially from `involutive reverse` and `involutive (map reverse)`, and a proof of one of them.

Reply:
Should I stick to the applied versions of the lemmas, or use the `∘` forms we're discussing here?

Reply:
Applied should be fine. (We should really make a tactic that derives one set of lemmas from the other!)

Reply:
Will add these once the merge build finishes and I have a cache to work from

Reply:
Added the six lemmas you suggested, and tweaked the names to match other lemmas in the file